### DISCUSSION TITLE
Flagging and highlighting flows automatically from list of regexes

### DISCUSSION THREAD

**Discussion by malwareinfosec** [mitmproxy/mitmproxy#6699] [Ideas]
Hello,

I am the author of EKFiddle ([https://github.com/malwareinfosec/EKFiddle](https://github.com/malwareinfosec/EKFiddle)), a plugin for Fiddler. The primary use case is to detect malicious web traffic using regular expressions that apply to URI, IP, Response Body, etc... Any detection is shown to the user by changing the UI, i.e. adding colors to the web sessions and a description of the threat.

I've also done something similar for ZAP proxy ([https://github.com/malwareinfosec/FiddleZAP](https://github.com/malwareinfosec/FiddleZAP)).

I am interested to see if this is something that mitmproxy could support?

**Comment by Prinzhorn** [comment]
You can "mark" flows (https://docs.mitmproxy.org/stable/api/mitmproxy/flow.html#Flow.marked) and you can add `flow.comment` / `flow.metadata`, which shows up in the UI. If you would need anything more specific than that, it'd be better to open a feature request.

We'd be more than happy to include such an addon in the contrib folder https://github.com/mitmproxy/mitmproxy/tree/main/examples/contrib, not sure about shipping it. This is a very niche use case for mitmproxy I would assume.

**Reply by malwareinfosec** [reply]
Thanks for the quick reply!

Is it correct to assume that flow.comment/flow.metadata can be added to the UI via an external Python script? Does this happen in real time as flows are being recorded, or can it be done only on pre-loaded flows (and looping through them)?

I will also look into those community examples to get some ideas.

**Reply by Prinzhorn** [reply]
fiddle.py

```py
def request(flow):
    if flow.request.path == "/evil":
        flow.marked = ":red_circle:"
        flow.comment = "Requested path was /evil"
```

```sh
mitmproxy --scripts fiddle.py
curl --insecure --proxy localhost:8080 https://example.com/
curl --insecure --proxy localhost:8080 https://example.com/evil
```

![Screenshot from 2024-03-01 17-58-06](https://github.com/mitmproxy/mitmproxy/assets/679144/d0a8ca7b-d5f1-4141-a417-b8faa5be10ff)
![Screenshot from 2024-03-01 17-58-11](https://github.com/mitmproxy/mitmproxy/assets/679144/85f41848-9029-4ecf-8d91-2ca46fbf6c07)

**Reply by malwareinfosec** [reply]
Thank you, this is very straightforward.

I was able to do the same for mitmweb:

![image](https://github.com/mitmproxy/mitmproxy/assets/25351665/3b7d6c8d-9075-4f16-a71e-259dae21c7ac)

However, I do not see a way to add a 'comment' column within the UI. I tried to add it from the web_columns option but that didn't work. If there is a way, how do you make it permanent so that restarting mitmweb brings up the new column automatically?

![image](https://github.com/mitmproxy/mitmproxy/assets/25351665/3ff6f45c-362a-4d01-86f5-ea307db007cd)

**Reply by Prinzhorn** [reply]
I'm really not familiar with mitmweb, but it looks like the data is correctly send to the client

![image](https://github.com/mitmproxy/mitmproxy/assets/679144/b37c31b9-7a6a-4281-81eb-46aa5af35072)

so it might be a matter of adding `comments` to the list of allowed columns or we should render it in the flow details on the right. Actually both probably, but in the column it would just be a single truncated line with overflow ellipsis.

If you have any interest in submitting a PR it seems we export the different columns here https://github.com/mitmproxy/mitmproxy/blob/main/web/src/js/components/FlowTable/FlowColumns.tsx and comments would be filtered here because of missing implementation https://github.com/mitmproxy/mitmproxy/blob/590eef0d44d0c052b6f67609077993e7655bf4f0/web/src/js/components/FlowTable/FlowRow.tsx#L46

**Reply by malwareinfosec** [reply]
Having comments in both the columns and flow details would be great. I'm not sure how to do this kind of pull request though.

I've created an early version script based on your original code snippet that checks _flow.request.path_ and _flow.response.content_ against a list of regexes:
[https://github.com/malwareinfosec/fiddleitm](https://github.com/malwareinfosec/fiddleitm)

**Reply by Prinzhorn** [reply]
I've opened https://github.com/mitmproxy/mitmproxy/issues/6708

Great!

**Comment by lups2000** [comment]
I opened a new PR for this, confirm me that is what you wanted. Otherwise I can change it , thanks :)
